### PR TITLE
Adding support for a IUserClaimStore along with Claim to attribute mapping

### PR DESCRIPTION
**Pull request by assyadh**
*Issue #, if available:* DOTNET-3057

*Description of changes:*

Adding support for a IUserClaimStore along with Claim to attribute mapping

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs [hunk 1]
```diff
@@ -59,5 +61,44 @@ public async Task<ClaimsPrincipal> CreateAsync(TUser user)
             var claimsIdentity = new ClaimsIdentity(claims, IdentityConstants.ApplicationScheme);
             return new ClaimsPrincipal(claimsIdentity);
         }
+
+        /// <summary>
+        /// Internal method to map System.Security.Claims.ClaimTypes to Cognito Standard Attributes
+        /// </summary>
+        /// <param name="claims"></param>
+        private void MapClaimTypesToCognito(List<Claim> claims)
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs]: IEnumerable parameter

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs [hunk 2]
```diff
@@ -59,5 +61,44 @@ public async Task<ClaimsPrincipal> CreateAsync(TUser user)
...
+        {
+            var emailClaim = claims.FirstOrDefault(claim => claim.Type == CognitoAttributesConstants.Email);
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs]: can we add each of these 3 line null/add checks in a method that take a claim and a constant, and return a Claim/null?

Also, we can just add the the list blindly and remove all nulls at the end.

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserManager.cs [hunk 3]
```diff
@@ -445,5 +448,55 @@ protected override Task<IdentityResult> UpdateUserAsync(TUser user)
...
+        /// <param name="user">The user to add the claim to.</param>
+        /// <param name="claims">The claims to add.</param>
+        /// <returns>
+        /// The <see cref="Task"/> that represents the asynchronous operation, containing the <see cref="IdentityResult"/>
+        /// of the operation.
+        /// </returns>
+        public override async Task<IdentityResult> AddClaimsAsync(TUser user, IEnumerable<Claim> claims)
+        {
+            ThrowIfDisposed();
+            if (user == null)
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoUserManager.cs]: Checking for user and claims to not be null is a common check. Can we make a static method to delegate to? Heck, you could even make it generic by having it take in a list of tuples (name, value).

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs (hunk 1 above)
Comment (assyadh) [reply src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs]: Done

Comment (assyadh) [reply src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs]: I need it to be a List to access the .Add() method.

I think this is OK as it is a private utility method.

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs (hunk 2 above)
Comment (assyadh) [reply src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs]: Moved the mapping to a Dictionary<string, string>.

Extracted the mapping logic to a small function.

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserManager.cs (hunk 3 above)
Comment (assyadh) [reply src/Amazon.AspNetCore.Identity.Cognito/CognitoUserManager.cs]: I would rather keep it this way for consistency with the other Identity providers:

https://github.com/aspnet/Identity/blob/eb3ff7fc32dbfff65a1ba6dfdca16487e0f6fc41/src/Microsoft.Extensions.Identity.Core/UserManager.cs#L1027

https://github.com/JudahGabriel/RavenDB.Identity/blob/e898bf6f0b33471390a8e75ce828bd4f7a305873/RavenDB.Identity/RoleStore.cs#L128

While I agree the checks may seem redundant, our codebase is already disgressing from Microsoft signatures due to the specificities of Cognito.

File: src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs [hunk 4]
```diff
@@ -29,6 +29,18 @@ public class CognitoUserClaimsPrincipalFactory<TUser> : IUserClaimsPrincipalFact
         private readonly CognitoUserManager<TUser> _userManager;
         private readonly IdentityOptions _identityOptions;
 
+        private Dictionary<string, string> claimToAttributesMapping = new Dictionary<string, string>()
```
Comment (costleya) [inline src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs]: readonly

Comment (assyadh) [reply src/Amazon.AspNetCore.Identity.Cognito/CognitoUserClaimsPrincipalFactory.cs]: Done

**Pull request merged**